src/crimson/os/seastore/btree/fixed_kv_node.h

```diff
@@ -29,27 +29,453 @@ namespace crimson::os::seastore {
+   * 	c. copy_sources is empty
+   * 2. if nodes are mutation_pending:
+   * 	a. parent points at stable instance of the same node
+   * 	b. child pointers are null except for initial_pending() children of
+   * 	   this transaction. Child resolution is done by first checking this
+   * 	   array, and then recursively resolving via the parent. We copy child
+   * 	   pointers from parent on commit.
+   * 	c. copy_sources is empty
+   * 3. if nodes are initial_pending
+   * 	a. parent points at is_mutation_pending() parent on this transaction
+   * 	   (must exist)
+   * 	b. child pointers are null except for initial_pending() children of
+   * 	   this transaction (live due to 3a below). Child resolution is done
+   * 	   by first checking this array, and then recursively resolving via
+   * 	   the correct copy_sources entry. We copy child pointers from copy_sources
+   * 	   on commit.
+   * 	c. copy_sources contains the set of stable nodes with which the lba
+   * 	   range of this node overlaps.
+   */
+  std::vector<CachedExtent*> children;
+  std::set<FixedKVNodeRef, copy_source_cmp_t> copy_sources;
+  uint16_t capacity = 0;
+  parent_tracker_t* my_tracker = nullptr;
+  parent_tracker_ref parent_tracker;
+
+  FixedKVNode(uint16_t capacity, ceph::bufferptr &&ptr)
+    : CachedExtent(std::move(ptr)),
+      pin(this),
+      children(capacity, nullptr),
+      capacity(capacity) {}
   FixedKVNode(const FixedKVNode &rhs)
-    : CachedExtent(rhs), pin(rhs.pin, this) {}
+    : CachedExtent(rhs),
+      pin(rhs.pin, this),
+      children(rhs.capacity, nullptr),
+      capacity(rhs.capacity) {}
 
   virtual fixed_kv_node_meta_t<node_key_t> get_node_meta() const = 0;
+  virtual uint16_t get_node_size() const = 0;
 
   virtual ~FixedKVNode() = default;
+  virtual node_key_t get_key_from_idx(uint16_t idx) const = 0;
+
+  template<typename iter_t>
+  void update_child_ptr(iter_t iter, ChildableCachedExtent* child) {
+    children[iter.get_offset()] = child;
+    set_child_ptracker(child);
+  }
+
+  template<typename iter_t>
+  void insert_child_ptr(iter_t iter, ChildableCachedExtent* child) {
+    auto raw_children = children.data();
+    auto offset = iter.get_offset();
+    std::memmove(
+      &raw_children[offset + 1],
+      &raw_children[offset],
+      (get_node_size() - offset) * sizeof(ChildableCachedExtent*));
+    children[offset] = child;
+    set_child_ptracker(child);
+  }
+
+  template<typename iter_t>
+  void remove_child_ptr(iter_t iter) {
+    LOG_PREFIX(FixedKVNode::remove_child_ptr);
+    auto raw_children = children.data();
+    auto offset = iter.get_offset();
+    SUBTRACE(seastore_fixedkv_tree, "trans.{}, pos {}, total size {}, extent {}",
+      this->pending_for_transaction,
+      offset,
+      get_node_size(),
+      (void*)raw_children[offset]);
+    std::memmove(
+      &raw_children[offset],
+      &raw_children[offset + 1],
+      (get_node_size() - offset - 1) * sizeof(ChildableCachedExtent*));
```

Comment:
No, the actual removal happens on transaction commit, reset child's parent tracker here will cause error if the transaction is reset.

Reply:
This node is only possible to be mutation-pending or initial-pending, so it's children must be initial-pending from the design. If it is correct, why it will cause error to unlink the initial-pending child's parent-tracker?

Seems it won't cause real issues yet, but IIUC there is an invarient that the child-(ptracker)->parent reference must be consistent with parent->children ptr.

Reply:
> This node is only possible to be mutation-pending or initial-pending, so it's children must be initial-pending from the design. If it is correct, why it will cause error to unlink the initial-pending child's parent-tracker?

Ah, sorry, I miss understood you. When a pending child is removed, it'll be retired and its `on_invalidated()` will be called in which the parent tracker is reset.

Reply:
Well, correct. But I think it would be better to see the invariant is preserved whenever operate on the link. I think I'll still forget that it will be finally fixed in the on_invalidate() callback.

Or probably add a comment here, should be good as well.

Reply:
Yes, will do.